### PR TITLE
feat: add kicker to teaser

### PR DESCRIPTION
**Pull request by sitepark-becker**
Adds a kicker text to `ArticleTeaser`s and `MediaTeaser`s. 

Kickers texts are resolved in the following way:
1. check the resource data for `base.teaser.kicker` 
2. if empty, check the resource data for `base.kicker` 
3. if empty, traverse navigation tree upwards until a resource with `base.kicker` is found

Step 3 is ignored for `MediaTeaser` as media articles aren't part of the navigation tree.

Currently missing:
- microsite mounting is currently ignored

### CODE REVIEW

**Comment by sitepark-veltrup** [comment]
`ArticleTeaser` and the `ArticleTeaserFactory` should not be adapted for this. I think it would be better if we extended the `ArticleTeaserResolver` for this. See [`getDate`](https://github.com/sitepark/atoolo-graphql-search-bundle/blob/feature/core-name/src/Resolver/ArticleTeaserResolver.php#L44) and `getAsset()`

This way, the method can also be used as in [`NewsTeaserResolver` ](https://github.com/sitepark/atoolo-graphql-search-bundle/blob/feature/core-name/src/Resolver/NewsTeaserResolver.php).

I think the lookup technique should also work for media teasers.

I think a kicker should not be a string but there should be a `Kicker` interface. There can be different kickers for articles, news, events, etc. that provide more than one string. With `Asset` this is also solved via an interface, you can take a look there. It is important that all interface implementations [here](https://github.com/sitepark/atoolo-graphql-search-bundle/blob/feature/core-name/src/Resources/config/graphql.yaml#L6) are added.

**Comment by sitepark-becker** [comment]
I moved the logic into `ArticleTeaserResolver::getKicker()` and removed the kicker from the `ArticleTeaser`-Type. That way media teasers will have the same lookup mechanism for now.

Im not so sure on the `Kicker` interface though. Sure, from a frontend point of view, it might be the case that teasers will render a date or some file info in their "kicker element", but that doesn't mean that the date/file info are inherently part of a kicker interface, does it? E.g. a `NewsTeaser` already has a `date` field and it's not necessary to duplicate this field in a kicker interface.

I think from a backend perspective, the word "kicker" should really only refer to the kicker text, a string.

**Comment by sitepark-veltrup** [comment]
I have cases where a kicker is more than a string, e.g. at news:

```html
<div class="SP-Kicker SP-Teaser__kicker">
  <span class="SP-Kicker__date">19.04.2024</span>
  <span class="SP-Kicker__text">Administration</span>
</div>
```
For events, it can be a list of categories. Or there is also a label such as 'sold out'.

There are many more cases.

**Comment by sitepark-veltrup** [comment]
I think the question is where the responsibility lies, who collects the data for a kicker. Does the backend do this or is it the responsibility of the frontend to collect all the necessary data to display it in the kicker. I can imagine that the backend still takes on this task.

**Comment by sitepark-becker** [comment]
> For events, it can be a list of categories. Or there is also a label such as 'sold out'.

I those cases, wouldn't it be easier to add the missing fields `categories` and `status` directly to a potential `EventTeaser` type instead of making it explicitly part of a kicker interface? Then the frontend could still decide to render it as part of the kicker if they want to. 
Cause as I said, in cases like this the `date` field would be redundant:
```json
{
    "__typename": "NewsTeaser",
    "kicker": {
        "text": "Kicker",
        "date": "01.01.2024"
    },
    "date": "01.01.2024"
}
```

**Comment by sitepark-veltrup** [comment]
Maybe we'll talk to the front end about what they think makes more sense. If we decide to use `String` now, we won't be able to get away from it.

**Comment by sitepark-veltrup** [comment]
> Step 3 is ignored for MediaTeaser as media articles aren't part of the navigation tree.

I don't think that's quite right. They are at least attached to the homepage. This allows media from microsites to have a different kicker and you can recognize that the media comes from a microsite.

But maybe this is irrelevant for media, because most of them have a media kicker with file name and file size.

File: src/Resolver/ArticleTeaserFactory.php [hunk 1]
```diff
@@ -32,6 +32,7 @@ public function create(Resource $resource): Teaser
             $url,
             $headline,
             $text,
+            null, // will be resolved by getAsset()
```
[the remark this answers is not in the document]
Comment (sitepark-becker) [reply src/Resolver/ArticleTeaserFactory.php]: Fixed

File: src/Resolver/ArticleTeaserResolver.php [hunk 2]
```diff
@@ -37,10 +38,40 @@ class ArticleTeaserResolver implements Resolver
...
+        $primaryPath = $this->navigationLoader->loadPrimaryPath(
+            $resource->toLocation()
+        );
+        for ($i = count($primaryPath) - 1; $i >= 1; $i--) {
+            $kicker = $primaryPath[$i]->data->getString('base.kicker');
+            if (!empty($kicker)) {
+                return $kicker;
+            }
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (sitepark-becker) [reply src/Resolver/ArticleTeaserResolver.php]: Looks good. I added a few lines to check the current resource first though.

**Comment by sitepark-becker** [comment]
> Maybe we'll talk to the front end about what they think makes more sense. If we decide to use `String` now, we won't be able to get away from it.

Hendrik said he's ok with `kicker` always referring to a nullable string. We don't have to merge these changes today though, I'm fine with waiting till monday

> I don't think that's quite right. They are at least attached to the homepage. This allows media from microsites to have a different kicker and you can recognize that the media comes from a microsite.
> 
> But maybe this is irrelevant for media, because most of them have a media kicker with file name and file size.

Since I refactored my changes the media teaser now use the same lookup mechanism as article teasers anyway. So yeah I think your right

**Comment by sitepark-veltrup** [comment]
> Hendrik said he's ok with kicker always referring to a nullable string. We don't have to merge these changes today though, I'm fine with waiting till monday

I've now thought about it again and I think you're right. We provide the data, the front end decides where and how it is displayed.

Your changes have pulled down the code coverage. There are still tests missing.

https://github.com/sitepark/atoolo-graphql-search-bundle/pull/37#issuecomment-2066259741

**Comment by sitepark-becker** [comment]
Code should be covered now (even thought the codecov bot doesn't seem to acknowledge?)

File: test/Resolver/ArticleTeaserResolverTest.php [hunk 3]
```diff
@@ -61,6 +88,58 @@ public function testGetDate(): void
...
+            $this->resolver->getKicker($teaserC),
+            'unexpected teaser kicker'
+        );
+        $this->assertEquals(
+            null,
+            $this->resolver->getKicker($teaserD),
+            'unexpected teaser kicker'
+        );
+    }
+
```
Comment (sitepark-veltrup) [inline test/Resolver/ArticleTeaserResolverTest.php:142]: That's three tests in one. They should be divided into 3 tests.

Comment (sitepark-becker) [reply test/Resolver/ArticleTeaserResolverTest.php]: Refactored it to have one assert per test only

**Pull request merged**